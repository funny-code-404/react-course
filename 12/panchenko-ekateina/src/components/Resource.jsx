import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import styled from "styled-components";

const List = styled.ul`
  display: grid;
  grid-template-columns: 1fr 1fr;
  margin: auto;
  padding: 30px;
  width: 80%;

  @media (max-width: 590px) {
    grid-template-columns: 1fr;
    text-align: center;
  }
`;

const Item = styled.li`
  font-size: 20px;
  font-size: 25px;
  list-style-type: none;
  text-transform: uppercase;
  padding-bottom: 10px;
`;

const LinkItem = styled(Link)`
  color: #fff;
  cursor: pointer;
  margin-right: 10px;
  text-decoration: none;
`;

export const Resource = ({ url, action, data, type }) => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (url && !data) {
      dispatch(action(url));
    }
  }, [url, data]);

  return (
    <List>
      {data?.map((item) => (
        <Item key={item.id}>
          <LinkItem
            to={{
              pathname: `/${type}/${item.name.toLowerCase()}`,
              state: {
                data: item,
              },
            }}
          >
            {item.name}
          </LinkItem>
        </Item>
      ))}
    </List>
  );
};
