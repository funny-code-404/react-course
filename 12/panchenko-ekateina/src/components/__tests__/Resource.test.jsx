import { configure, mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { Resource } from "../Resource";
import "jest-styled-components";

configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  Link: "a",
  Route: ({ children, path }) => children({ match: path === "/" }),
}));
jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
}));

describe("Resource mounting behavior", () => {
  const props = {
    url: "/test",
    action: jest.fn(),
    data: [{ id: 1, name: "test" }],
    type: "test",
  };
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<Resource {...props} />);
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render 1 component snapshot", () => {
    expect(wrapper).toHaveLength(1);
  });

  it("should render 'Resource' component with actual prop", () => {
    expect(wrapper.find("a")).toHaveLength(1);
    expect(wrapper.find("a").props().to).toEqual({
      pathname: "/test/test",
      state: {
        data: {
          id: 1,
          name: "test",
        },
      },
    });
  });
});
