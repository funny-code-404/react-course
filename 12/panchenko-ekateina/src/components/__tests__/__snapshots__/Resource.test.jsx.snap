// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource mounting behavior should match snapshot 1`] = `
.c0 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  margin: auto;
  padding: 30px;
  width: 80%;
}

.c1 {
  font-size: 20px;
  font-size: 25px;
  list-style-type: none;
  text-transform: uppercase;
  padding-bottom: 10px;
}

.c2 {
  color: #fff;
  cursor: pointer;
  margin-right: 10px;
  -webkit-text-decoration: none;
  text-decoration: none;
}

@media (max-width:590px) {
  .c0 {
    grid-template-columns: 1fr;
    text-align: center;
  }
}

<Resource
  action={[MockFunction]}
  data={
    Array [
      Object {
        "id": 1,
        "name": "test",
      },
    ]
  }
  type="test"
  url="/test"
>
  <styled.ul>
    <ul
      className="c0"
    >
      <styled.li
        key="1"
      >
        <li
          className="c1"
        >
          <styled.a
            to={
              Object {
                "pathname": "/test/test",
                "state": Object {
                  "data": Object {
                    "id": 1,
                    "name": "test",
                  },
                },
              }
            }
          >
            <a
              className="c2"
              to={
                Object {
                  "pathname": "/test/test",
                  "state": Object {
                    "data": Object {
                      "id": 1,
                      "name": "test",
                    },
                  },
                }
              }
            >
              test
            </a>
          </styled.a>
        </li>
      </styled.li>
    </ul>
  </styled.ul>
</Resource>
`;
